### ISSUE TITLE
[BUG] Breaking changes for CCR plugin in OS 2.0

### ISSUE THREAD

**Issue by saikaranam-amazon** [opensearch-project/OpenSearch#2482]
**Describe the bug**
CCR depends on IndexShard(getHistoryOperations) to retrieve translog operations for leader index.
In 2.0, this method is removed in this PR: https://github.com/opensearch-project/OpenSearch/pull/2077 

```
 public Translog.Snapshot getHistoryOperations(String reason, Engine.HistorySource source, long startingSeqNo, long endSeqNo)
```

**Expected behavior**
- Please provide replacement for this method in 2.0 to retrieve operations based on history-source.

**Comment by dblock** [comment]
@saratvemulapalli What was this deprecated for?

**Comment by saratvemulapalli** [comment]
@saikaranam-amazon the change was intentional because soft deletes are not optional anymore and history doesnt exist in translog instead has to be fetched from lucene index. 

Could you try exploring the option of using:
  ```
public Translog.Snapshot getHistoryOperations(String reason, long startingSeqNo, long endSeqNo) throws IOException {
```
Please see the conversation history on: https://github.com/opensearch-project/OpenSearch/pull/2077#issuecomment-1071105592

Give this a shot and let us know how it goes. I'll keep the issue open

**Comment by saikaranam-amazon** [comment]
@saratvemulapalli Please this issue for context: https://github.com/opensearch-project/OpenSearch/issues/1100
We used this method to fetch the ops from Tlog as source and we can't fetch from the lucene index (due to reasons in the above issue).

**Comment by saikaranam-amazon** [comment]
> are not optional anymore and history doesnt exist in translog instead has to be fetched from lucene index.

Yes, previously as well that used to be the case. We have modified to hold the translog. This method is not alternative to the previous one as they don't give the same performance.

**Comment by saikaranam-amazon** [comment]
Gentle reminder! - Any update?

**Comment by dblock** [comment]
@saikaranam-amazon Does the option suggested above by @saratvemulapalli not work?

**Comment by saikaranam-amazon** [comment]
No @dblock , we will need the previous method due to the above reason.

**Comment by dblock** [comment]
@saikaranam-amazon What do you propose? 

@andrross @kartg Could I get you two to pitch in here, what should we do?

**Comment by saratvemulapalli** [comment]
@saikaranam-amazon from what I understand looks like we are writing to translog anymore. 
Adding back the method would not help. Again this is with my limited knowledge. 

From what you are saying its definitely an impact in performance when using lucene indices. But it feels like thats the only way. 
@nknize what would suggest?

**Comment by saratvemulapalli** [comment]
@saikaranam-amazon also in terms of the numbers, do you happen to know how bad it would be to read from lucene index vs reading from translog?

**Comment by nknize** [comment]
> due to Lucene stored fields decompression.


What version of Lucene did you benchmark? 

For clarification, operations are still available and can be replayed. It's just stored in a Lucene index instead of a Translog file. Hence "HistorySource" being removed as an option.

**Comment by nknize** [comment]
I just checked the commit. You were [benchmarking with Lucene 8.9](https://github.com/opensearch-project/OpenSearch/blob/a9e1913bd5c9ab4906e9cac2c189ed5f797a2a53/buildSrc/version.properties#L2) which had a [performance regression bug](https://issues.apache.org/jira/plugins/servlet/mobile#issue/LUCENE-9486) for Stored fields decompression that was [fixed in Lucene 8.10](https://issues.apache.org/jira/plugins/servlet/mobile#issue/LUCENE-9917)

**Comment by saikaranam-amazon** [comment]
Ack. Let me check and run with 8.10

**Comment by saikaranam-amazon** [comment]
For 2.0: Ran with latest 2.0.0 (alpha) artifacts(April 6th 2022:
For 1.3: Picked the latest release artifacts


Instance: i3.xlarge
Esrally dataset: noaa (Global daily weather measurements from NOAA )


Test setup:
 - In order to log only replication specific operations
    - Ensured single batch can be made during the following phase 
    - CPU and total time are captured.

Seeing a regression of ~10%(avg) diff and
difference of total time > ~15min to fetch the same amount of operations in 2.0 code path

**Comment by nknize** [comment]
> Esrally dataset: noaa (Global daily weather measurements from NOAA )

are you using the OpenSearch default codec or are you relying on rally to set the codec for you?

**Comment by saikaranam-amazon** [comment]
No, ESRally didn't set the codec. 

The index is created with the default codec.

**Comment by nknize** [comment]
> The index is created with the default codec.

Can you link to the line where you're setting the codec?

**Comment by saikaranam-amazon** [comment]
No, I didn't set any codec for the index. It is the default codec that every index picks on new index creation.

**Comment by nknize** [comment]
Just a couple thoughts here.

1. Let's explicitly set `"index.codec" : "DEFAULT"` just rule out any possibilities of this measure the `BEST_SPEED` codec. 
2. We need to get a bigger picture of the benchmarks. Let's get comparison of disk usage when using translog vs lucene index as history source to show the trade of disk storage cost vs cpu performance.
3. Let's run with a few different [data sets](https://github.com/elastic/rally-tracks) to see if the regression is consistent across the board and not a red herring to NOAA. Can you start w/ the event data from the [original post](https://github.com/opensearch-project/OpenSearch/issues/1100#issuecomment-918290182). Then try nyc_taxi, so, noaa, nested?.
4. Can you include the benchmark table like included in the [original issue](https://github.com/opensearch-project/OpenSearch/issues/1100#issuecomment-918290182)?

**Comment by nknize** [comment]
Just touching base on this issue. The best approach for the project at this juncture is to leave as is; especially since 2.0 preview is right around the corner. 

I then suggest:

1. More diligent benchmarking to assess the speed vs storage tradeoffs
2. Further review to verify if this is due to the lucene compression changes
3. Discuss options
    a. explore making the [block size](https://issues.apache.org/jira/browse/LUCENE-9917) a dynamic `expert` setting
    b. explore adding a user setting to CCR to allow explicit speed (TLog) vs storage (LuceneSnapshot) (e.g., possibly use `index.codec`?)
  
Either way, if we need to add the TLog history source back, then that code logic should be encapsulated to the CCR plugin since it is no longer used in the core. Extension points can be added to the core similar to the way we added `EngineConfig` extensions to enable the retention lease Translog Deletion policy setting in 1.x.

**Comment by saikaranam-amazon** [comment]
> LuceneChangesSnapshot already retrieves the ops sorted by sequence number. Why all the transient ArrayLists and extra logic? Seems unnecessary.

This was used for the Tlog logic and not the new method from the OS core. 
Updated the method to remove any inconsistencies to run the tests.


> Just touching base on this issue. The best approach for the project at this juncture is to leave as is; especially since 2.0 preview is right around the corner.

I would like to add another option to add the methods back to the core package for 2.0 and have the path for deprecation in later 2.x release.


Regarding the previous runs and various datasets used:
 - Index was set with default codec which is tuned for best speed
 - Attaching the results for the noaa with esrally metrics. we have other datasets that are running in parallel and will update the issue as and when the results are available.
 - we already have event data(from previous run in the linked issue) and NOAA(in the current issue) exhibiting similar behavior.
   - In both of the cases, it is tuned for best_speed and not for compression(deflate codec).



Setup is same as before
 - 1 Primary 0 replicas
 - Index setting for codec set of default 
 - Started the replication on the indices and caputed the esrally metrics below



* 2.0 
```

------------------------------------------------------
    _______             __   _____
   / ____(_)___  ____ _/ /  / ___/_________  ________
  / /_  / / __ \/ __ `/ /   \__ \/ ___/ __ \/ ___/ _ \
 / __/ / / / / / /_/ / /   ___/ / /__/ /_/ / /  /  __/
/_/   /_/_/ /_/\__,_/_/   /____/\___/\____/_/   \___/
------------------------------------------------------
            
|                                                         Metric |   Task |    Value |   Unit |
|---------------------------------------------------------------:|-------:|---------:|-------:|
|                     Cumulative indexing time of primary shards |        |  89.2113 |    min |
|             Min cumulative indexing time across primary shards |        |        0 |    min |
|          Median cumulative indexing time across primary shards |        |  44.6057 |    min |
|             Max cumulative indexing time across primary shards |        |  89.2113 |    min |
|            Cumulative indexing throttle time of primary shards |        |        0 |    min |
|    Min cumulative indexing throttle time across primary shards |        |        0 |    min |
| Median cumulative indexing throttle time across primary shards |        |        0 |    min |
|    Max cumulative indexing throttle time across primary shards |        |        0 |    min |
|                        Cumulative merge time of primary shards |        |  18.1808 |    min |
|                       Cumulative merge count of primary shards |        |       28 |        |
|                Min cumulative merge time across primary shards |        |        0 |    min |
|             Median cumulative merge time across primary shards |        |   9.0904 |    min |
|                Max cumulative merge time across primary shards |        |  18.1808 |    min |
|               Cumulative merge throttle time of primary shards |        |  7.59582 |    min |
|       Min cumulative merge throttle time across primary shards |        |        0 |    min |
|    Median cumulative merge throttle time across primary shards |        |  3.79791 |    min |
|       Max cumulative merge throttle time across primary shards |        |  7.59582 |    min |
|                      Cumulative refresh time of primary shards |        |  1.24653 |    min |
|                     Cumulative refresh count of primary shards |        |       54 |        |
|              Min cumulative refresh time across primary shards |        |        0 |    min |
|           Median cumulative refresh time across primary shards |        | 0.623267 |    min |
|              Max cumulative refresh time across primary shards |        |  1.24653 |    min |
|                        Cumulative flush time of primary shards |        |  4.16113 |    min |
|                       Cumulative flush count of primary shards |        |       26 |        |
|                Min cumulative flush time across primary shards |        |        0 |    min |
|             Median cumulative flush time across primary shards |        |  2.08057 |    min |
|                Max cumulative flush time across primary shards |        |  4.16113 |    min |
|                                             Total Young Gen GC |        |  194.223 |      s |
|                                               Total Old Gen GC |        |    0.305 |      s |
|                                                     Store size |        |  7.29137 |     GB |
|                                                  Translog size |        | 0.524542 |     GB |
|                                         Heap used for segments |        |        0 |     MB |
|                                       Heap used for doc values |        |        0 |     MB |
|                                            Heap used for terms |        |        0 |     MB |
|                                            Heap used for norms |        |        0 |     MB |
|                                           Heap used for points |        |        0 |     MB |
|                                    Heap used for stored fields |        |        0 |     MB |
|                                                  Segment count |        |       72 |        |
|                                                 Min Throughput |  index |    23000 | docs/s |
|                                              Median Throughput |  index |  23436.3 | docs/s |
|                                                 Max Throughput |  index |  24577.6 | docs/s |
|                                        50th percentile latency |  index |  1599.55 |     ms |
|                                        90th percentile latency |  index |  1986.58 |     ms |
|                                        99th percentile latency |  index |  4399.13 |     ms |
|                                      99.9th percentile latency |  index |    10353 |     ms |
|                                       100th percentile latency |  index |  13417.4 |     ms |
|                                   50th percentile service time |  index |  1599.55 |     ms |
|                                   90th percentile service time |  index |  1986.58 |     ms |
|                                   99th percentile service time |  index |  4399.13 |     ms |
|                                 99.9th percentile service time |  index |    10353 |     ms |
|                                  100th percentile service time |  index |  13417.4 |     ms |
|                                                     error rate |  index |        0 |      % |


----------------------------------
[INFO] SUCCESS (took 1473 seconds)
----------------------------------

```

* 1.3
```
------------------------------------------------------
    _______             __   _____
   / ____(_)___  ____ _/ /  / ___/_________  ________
  / /_  / / __ \/ __ `/ /   \__ \/ ___/ __ \/ ___/ _ \
 / __/ / / / / / /_/ / /   ___/ / /__/ /_/ / /  /  __/
/_/   /_/_/ /_/\__,_/_/   /____/\___/\____/_/   \___/
------------------------------------------------------
            
|                                                         Metric |   Task |      Value |   Unit |
|---------------------------------------------------------------:|-------:|-----------:|-------:|
|                     Cumulative indexing time of primary shards |        |    75.3275 |    min |
|             Min cumulative indexing time across primary shards |        |          0 |    min |
|          Median cumulative indexing time across primary shards |        |    37.6638 |    min |
|             Max cumulative indexing time across primary shards |        |    75.3275 |    min |
|            Cumulative indexing throttle time of primary shards |        |          0 |    min |
|    Min cumulative indexing throttle time across primary shards |        |          0 |    min |
| Median cumulative indexing throttle time across primary shards |        |          0 |    min |
|    Max cumulative indexing throttle time across primary shards |        |          0 |    min |
|                        Cumulative merge time of primary shards |        |    17.3825 |    min |
|                       Cumulative merge count of primary shards |        |         26 |        |
|                Min cumulative merge time across primary shards |        |          0 |    min |
|             Median cumulative merge time across primary shards |        |    8.69126 |    min |
|                Max cumulative merge time across primary shards |        |    17.3825 |    min |
|               Cumulative merge throttle time of primary shards |        |    6.83275 |    min |
|       Min cumulative merge throttle time across primary shards |        |          0 |    min |
|    Median cumulative merge throttle time across primary shards |        |    3.41637 |    min |
|       Max cumulative merge throttle time across primary shards |        |    6.83275 |    min |
|                      Cumulative refresh time of primary shards |        |   0.800183 |    min |
|                     Cumulative refresh count of primary shards |        |         46 |        |
|              Min cumulative refresh time across primary shards |        |          0 |    min |
|           Median cumulative refresh time across primary shards |        |   0.400092 |    min |
|              Max cumulative refresh time across primary shards |        |   0.800183 |    min |
|                        Cumulative flush time of primary shards |        |    4.12702 |    min |
|                       Cumulative flush count of primary shards |        |         27 |        |
|                Min cumulative flush time across primary shards |        |          0 |    min |
|             Median cumulative flush time across primary shards |        |    2.06351 |    min |
|                Max cumulative flush time across primary shards |        |    4.12702 |    min |
|                                             Total Young Gen GC |        |    149.146 |      s |
|                                               Total Old Gen GC |        |      0.224 |      s |
|                                                     Store size |        |    6.72589 |     GB |
|                                                  Translog size |        |   0.522197 |     GB |
|                                         Heap used for segments |        |    1.21087 |     MB |
|                                       Heap used for doc values |        |   0.450932 |     MB |
|                                            Heap used for terms |        |   0.718445 |     MB |
|                                            Heap used for norms |        | 0.00012207 |     MB |
|                                           Heap used for points |        |          0 |     MB |
|                                    Heap used for stored fields |        |  0.0413742 |     MB |
|                                                  Segment count |        |         67 |        |
|                                                 Min Throughput |  index |    21333.4 | docs/s |
|                                              Median Throughput |  index |    26861.7 | docs/s |
|                                                 Max Throughput |  index |    27170.9 | docs/s |
|                                        50th percentile latency |  index |    1301.35 |     ms |
|                                        90th percentile latency |  index |    1727.66 |     ms |
|                                        99th percentile latency |  index |    4972.26 |     ms |
|                                      99.9th percentile latency |  index |    9734.46 |     ms |
|                                       100th percentile latency |  index |    10881.1 |     ms |
|                                   50th percentile service time |  index |    1301.35 |     ms |
|                                   90th percentile service time |  index |    1727.66 |     ms |
|                                   99th percentile service time |  index |    4972.26 |     ms |
|                                 99.9th percentile service time |  index |    9734.46 |     ms |
|                                  100th percentile service time |  index |    10881.1 |     ms |
|                                                     error rate |  index |          0 |      % |


----------------------------------
[INFO] SUCCESS (took 1263 seconds)
----------------------------------
```

**Comment by nknize** [comment]
Thank you for posting this @saikaranam-amazon 

A few quick questions:

1. Did you index the same number of documents in both runs?
2. For the [event data benchmark](https://github.com/opensearch-project/OpenSearch/issues/1100#issuecomment-918290182) did you explicitly set the codec or leave it alone per the run?


> add the methods back to the core package for 2.0 and have the path for deprecation in later 2.x release.

Since the core doesn't use the methods it would be better to add as an Engine extension in the plugin.

**Comment by nknize** [comment]
> add the methods back to the core package for 2.0 and have the path for deprecation in later 2.x release.

Since public preview is right around the corner and CCR will need to migrate to [segrep](https://github.com/opensearch-project/OpenSearch/issues/2229) for 3.0+ anyway, how about a middle ground for 2.0+:

1. Open the PR to [add back deprecated TLog methods](https://github.com/nknize/OpenSearch/commit/ff6e60137735f7aabfab3d3415e7b45924ee1c2d) for **2.0 release only**
2. [Refactor CCR as a core module/plugin](https://github.com/opensearch-project/OpenSearch/issues/2872) for 2.1+
3. Dig into benchmarks to ensure validity
    a. based on benchmark findings either i. refactor deprecated methods to the new CCR core module/plugin or ii. remove altogether back .
4. Migrate CCR to segrep in 3.0+

Feel free to thumb up #2872 if agree

**Comment by mikemccand** [comment]
Sorry, I am new to CCR implementation, but my biggest question is why would we rely on replaying indexing operations from `xlog` to achieve CCR?  Shouldn't we rather replicate delta (changed since last CCR sync) Lucene's segments, which already did the costly work of indexing all docs?

**Comment by dblock** [comment]
@saikaranam-amazon please also confirm that you agree with the entire plan above, especially https://github.com/opensearch-project/OpenSearch/issues/2872 (https://github.com/opensearch-project/OpenSearch/pull/2886#issuecomment-1098489504). If we don't do this I'd rather not reintroduce the deprecated methods and ship with the regression.

**Comment by CEHENKLE** [comment]
Concur with dB.   We do not want to add technical debt with out a plan to remove it.

**Comment by Pallavi-AWS** [comment]
@dblock @nknize Thanks for the interim workaround of https://github.com/opensearch-project/OpenSearch/pull/2886. We will have a plan in place to move away from deprecated method in OS 2.1 (refactoring core and plugin will need some design work). However, for merging CCR into core, we can explore options for OS 3.0 (java migration etc.) - we might have to settle on a hybrid model where we continue to have functionality like security in the replication plugin and move replication engine to core.

**Comment by krishna-ggk** [comment]
> Sorry, I am new to CCR implementation, but my biggest question is why would we rely on replaying indexing operations from xlog to achieve CCR? Shouldn't we rather replicate delta (changed since last CCR sync) Lucene's segments, which already did the costly work of indexing all docs?

Thanks for the feedback Mike. There were a bunch of considerations for CCR to rely on translog at the time.

1. Ability to filter out documents to be replicated
2. Reliance on refreshes which may have implications to some usecases
3. Since segments on primary and replica are not homogeneous, replica promotions due to node losses etc will need special handling with more complexity.
4. Further due to same reason as 3, only a designated shard (primary/replica) can replicate segments - which means segment transfer cannot be load balanced if required.
5. Having different replication models for same cluster replicas vs different cluster (at the time) for the same index implies inheriting cons from both models.

That said, we totally realize the immense gains we can get with segment replication (thanks to your blogs as well as other success stories.) With [segment replication](https://github.com/opensearch-project/OpenSearch/issues/2229) making its way into OpenSearch for same-cluster replicas, it is a good time to visit CCR based on segrep.

**Comment by nknize** [comment]
> for merging CCR into core, we can explore options for OS 3.0... 

We need to do this in 2.x instead of opening more core extension points to override mechansims we shouldn't be overriding; we've already be burnt by that once when KNN and CCR were incompatible due to both overridding the Engine. 



> we might have to settle on a hybrid model where we continue to have functionality like security in the replication plugin and move replication engine to core.

As a first step we can refactor CCR to core and secure the endpoints separately in the security module. This fits with the design of core in 1.x/2.x and the secure endpoints will refactor to core once security becomes a first class citizen. 


> it is a good time to visit [CCR based on segrep](https://github.com/opensearch-project/OpenSearch/issues/2872).

:100:  CCR should be baked into the design of segrep from the beginning /cc @mch2 @kartg @anasalkouz @mikemccand

**Comment by nknize** [comment]
> Ability to filter out documents to be replicated

Just dropping some thoughts out loud on this one:

This sounds like a good use-case fit for the streaming index api (future home for the xlog and user defined durability policies). One idea is to have CCR filters the streaming index api route copies of indexing requests to the follower cluster if filter is matched. Another idea, there could be co-located IndexWriters in the streaming index api of the leader cluster that filter the documents and flush follower segments to be replicated by segrep of the follower cluster only.

In the former case you're using a form of docrep for inter-cluster replication but paying the penalty for additional network traffic (and possible failures). In the latter you're parallelizing the writes on the leader node and using segrep for inter and intra-cluster replication.

**Comment by dblock** [comment]
@saikaranam-amazon Care to add a summary with links to various PRs and close?

**Comment by mikemccand** [comment]
> > Sorry, I am new to CCR implementation, but my biggest question is why would we rely on replaying indexing operations from xlog to achieve CCR? Shouldn't we rather replicate delta (changed since last CCR sync) Lucene's segments, which already did the costly work of indexing all docs?
> 
> Thanks for the feedback Mike. There were a bunch of considerations for CCR to rely on translog at the time.

+1, thanks for the details. More replies below:

> 1. Ability to filter out documents to be replicated

Lucene has a powerful `FilterLeafReader` that works really well for this.  You can far more efficiently filter on the post-indexed form, segment by segment, instead of the raw original input source docs.

> 2. Reliance on refreshes which may have implications to some usecases

That's a good point -- if you somehow need CCR more frequently than refreshes than replicating segments might be an issue.  But does this really happen in practice?

> 3. Since segments on primary and replica are not homogeneous, replica promotions due to node losses etc will need special handling with more complexity.

Actually this is a big benefit of NRT segment replication -- the primary and replica ARE homogeneous.  If replica is promoted, due to either primary node down, or simple load balancing freedom in the cluster (master can pick a different replica that has colder CPU than the hot primary to protect overall cluster throughput), it's the exact same segments.  This is the same reason why scan/scroll can also be safely routed to any primary or its replicas with no loss of accuracy.

> 4. Further due to same reason as 3, only a designated shard (primary/replica) can replicate segments - which means segment transfer cannot be load balanced if required.

NRT segment replication fixes that too -- it can load balance nearly arbitrarily across primary + its replicas on every "pull next bytes" request.  Just needs the right lease logic to keep the point-in-time snapshots lit.

> 5. Having different replication models for same cluster replicas vs different cluster (at the time) for the same index implies inheriting cons from both models.

I would rather we deprecate the inefficient and fragile "reindex all docs" approach and encourage users to switch to NRT segment replication.  I agree keeping both options long-term is a bad choice.
 
> That said, we totally realize the immense gains we can get with segment replication (thanks to your blogs as well as other success stories.) With [segment replication](https://github.com/opensearch-project/OpenSearch/issues/2229) making its way into OpenSearch for same-cluster replicas, it is a good time to visit [CCR based on segrep](https://github.com/opensearch-project/OpenSearch/issues/2872).

Thanks.  I've long felt (decade at least) that it was crazy that ES chose document replication.  Lucene is naturally write-once at the file level and that makes for an obvious replication unit.  And it yields awesome transactional semantics to users, point-in-time searching and efficient incremental backups, like ZFS's choice of write-once block level storage yielding efficient whole-filesystem snapshots to users.

**Comment by dblock** [comment]
@Pallavi-AWS ^

**Comment by mikemccand** [comment]
> > Lucene has a powerful `FilterLeafReader` that works really well for this.
> 
> This is true! I wonder then, w/ segment replication is there even a need for CCR? Couldn't we just snapshot in the leader cluster and restore in the follower and use FLR on the follower to filter unneeded docs?
> 
> Is there a use case I'm missing where CCR is needed at all if we're using a segment level replication design? Seems more appropriate for document level.

+1 to explore this.

Let the cloud storage layer (S3 or whatever) handle the geo-distribution of the bits.  Snapshot in one geo and restore in the other.

**Comment by krishna-ggk** [comment]
> > it is a good time to visit [CCR based on segrep](https://github.com/opensearch-project/OpenSearch/issues/2872).
> 
> 💯 We can implement in a later phase but CCR should be baked into the design consideration of segrep from the beginning to minimize the early tech debt surface area /cc @mch2 @kartg @anasalkouz @mikemccand

@nknize - +1 - opened a dedicated [github issue](https://github.com/opensearch-project/cross-cluster-replication/issues/373) to explore this in more detail. 


> > 1. Ability to filter out documents to be replicated
> 
> Lucene has a powerful `FilterLeafReader` that works really well for this. You can far more efficiently filter on the post-indexed form, segment by segment, instead of the raw original input source docs.

True, definitely worth considering. One of the reason for considering filtering at source while replicating was to provide a security posture where filtered out documents don't move out of cluster. However given that current CCR doesn't implement filtering yet, we could consider seeking input from community to see if there is real need for this usecase in the RFC.

> > 2. Reliance on refreshes which may have implications to some usecases
> 
> That's a good point -- if you somehow need CCR more frequently than refreshes than replicating segments might be an issue. But does this really happen in practice?

Since CCR targets Disaster Recovery usecases, the thought was to ensure ack'd writes are replicated as soon as possible.  

> > 3. Since segments on primary and replica are not homogeneous, replica promotions due to node losses etc will need special handling with more complexity.
> 
> Actually this is a big benefit of NRT segment replication -- the primary and replica ARE homogeneous. If replica is promoted, due to either primary node down, or simple load balancing freedom in the cluster (master can pick a different replica that has colder CPU than the hot primary to protect overall cluster throughput), it's the exact same segments. This is the same reason why scan/scroll can also be safely routed to any primary or its replicas with no loss of accuracy.
>
> > 4. Further due to same reason as 3, only a designated shard (primary/replica) can replicate segments - which means segment transfer cannot be load balanced if required.
> 
> NRT segment replication fixes that too -- it can load balance nearly arbitrarily across primary + its replicas on every "pull next bytes" request. Just needs the right lease logic to keep the point-in-time snapshots lit.
>

Yes agree. However since we didn't have segment replication implementation for same-cluster replica at the time, we were limited by the cons of non-homogeneous segments.

> > 5. Having different replication models for same cluster replicas vs different cluster (at the time) for the same index implies inheriting cons from both models.
> 
> I would rather we deprecate the inefficient and fragile "reindex all docs" approach and encourage users to switch to NRT segment replication. I agree keeping both options long-term is a bad choice.

+1 

> > That said, we totally realize the immense gains we can get with segment replication (thanks to your blogs as well as other success stories.) With [segment replication](https://github.com/opensearch-project/OpenSearch/issues/2229) making its way into OpenSearch for same-cluster replicas, it is a good time to visit [CCR based on segrep](https://github.com/opensearch-project/OpenSearch/issues/2872).
> 
> Thanks. I've long felt (decade at least) that it was crazy that ES chose document replication. Lucene is naturally write-once at the file level and that makes for an obvious replication unit. And it yields awesome transactional semantics to users, point-in-time searching and efficient incremental backups, like ZFS's choice of write-once block level storage yielding efficient whole-filesystem snapshots to users.

@mikemccand  Thanks for the great feedback again :)

**Comment by Pallavi-AWS** [comment]
hi team, as this issue was tracking 2.0 CCR support (which is resolved), can we go ahead and close this issue? We will continue to discuss core migration under https://github.com/opensearch-project/OpenSearch/issues/2872 and segrep under https://github.com/opensearch-project/cross-cluster-replication/issues/373. @saikaranam-amazon can we please open one more issue on deprecating getHistory in 2.1? Thanks all.

**Comment by krishna-ggk** [comment]
> > Ability to filter out documents to be replicated
> 
> Just dropping some thoughts out loud on this one:
> 
> This sounds like a good use-case fit for the streaming index api (future home for the xlog and user defined durability policies). One idea is to have CCR filters in the streaming index api route copies of indexing requests to the follower cluster if filter is matched. Another idea, there could be co-located IndexWriters in the streaming index api of the leader cluster that filter the documents and flush follower segments to be replicated by segrep of the follower cluster only.
> 
> In the former case you're using a form of docrep for inter-cluster replication but paying the penalty for additional network traffic (and possible failures). In the latter you're parallelizing the writes on the leader node and using segrep for inter and intra-cluster replication. Incidentally this also provides the "fail early" protection we have today by indexing in memory and only replicating when flushing.
> > > Lucene has a powerful `FilterLeafReader` that works really well for this.
> > 
> > 
> > This is true! I wonder then, w/ segment replication is there even a need for CCR? Couldn't we just snapshot in the leader cluster and restore in the follower and use FLR on the follower to filter unneeded docs?
> > Is there a use case I'm missing where CCR is needed at all if we're using a segment level replication design? Seems more appropriate for document level.
> 
> +1 to explore this.
> 
> Let the cloud storage layer (S3 or whatever) handle the geo-distribution of the bits. Snapshot in one geo and restore in the other.

@nknize @mikemccand Thanks. Nice discussion and thoughts - continuing the discussion further in [373](https://github.com/opensearch-project/cross-cluster-replication/issues/373).

**Comment by saikaranam-amazon** [comment]
> A few quick questions:
Did you index the same number of documents in both runs?
For the event data benchmark did you explicitly set the codec or leave it alone per the run?


Yes, we do have same documents across the runs. And we did set the codec that is optimized for speed in both runs.
We have already a task open and that is in the review phase to automate the performance runs. This should address 
all the concerns for the future runs.


> Refactor CCR as a core module/plugin for 2.1+ and refactor deprecated methods to the CCR module/plugin

Sure. Opened issue under CCR to track this: https://github.com/opensearch-project/cross-cluster-replication/issues/374

> Dig into benchmarks to ensure validity
> a. based on benchmark findings either i. keep deprecated methods in the new CCR core module/plugin for all of 2.x or ii. remove altogether in favor of LuceneChangesSnapshot.

Understood and based on previous runs and subsequent runs performed (for this issue). we do see performance impact and already opened issue to track the work for providing extension points (see above)

> Migrate CCR to segrep in 3.0+

The work and the analysis will be captured in https://github.com/opensearch-project/cross-cluster-replication/issues/373 and discussion can be continued in that issue.

**Comment by saikaranam-amazon** [comment]
> for merging CCR into core, we can explore options for OS 3.0...
> > We need to do this in 2.x instead of opening more core extension points to override mechansims we shouldn't be overriding; we've already be burnt by that once when KNN and CCR were incompatible due to both overridding the Engine.

> we might have to settle on a hybrid model where we continue to have functionality like security in the replication plugin and move replication engine to core.
> > As a first step we can refactor CCR to core and secure the endpoints separately in the security module. This fits with the design of core in 1.x/2.x and the secure endpoints will refactor to core once security becomes a first class citizen. 

Agree with eventual state for the CCR and we will address the plugin migration(to core as module) and segment replication in the up-coming releases, as migration to core will require lot of effort(kotlin/java migration, security module etc) and planning based on the already scheduled releases for 2.x. 

we will continue the discussion to move the plugin as module under core in https://github.com/opensearch-project/OpenSearch/issues/2872 and Segment replication under CCR under: https://github.com/opensearch-project/cross-cluster-replication/issues/373
For 2.x, we will explore the option of extension points under https://github.com/opensearch-project/cross-cluster-replication/issues/374

**Comment by saikaranam-amazon** [comment]
> @saikaranam-amazon Care to add a summary with links to various PRs and close?

For 2.0, we have added the deprecated translog operations under the engine and PR(https://github.com/opensearch-project/OpenSearch/pull/2886) is merged for 2.0. Based on the further discussion, we have following issues for different tracks:
- CCR via segment replication: https://github.com/opensearch-project/cross-cluster-replication/issues/373
- Migraion of CCR Plugin to core module: https://github.com/opensearch-project/OpenSearch/issues/2872
- Extension points for 2.x release: https://github.com/opensearch-project/cross-cluster-replication/issues/374


Please continue the relevant discussions in the above issues.

Closing this one.
